README.mkd

```diff
@@ -229,6 +229,41 @@ registers and then save `ra` and the appropriate number of registers from
 epilogue. They restore the saved registers, deallocate the stack space for the
 register, and then perform a return through the restored value of `ra`.
 
+## Conventions for vendor extension
+
+Custom instruction is an important part of RISC-V, offical ISA spec has reserved
+serveral encoding space for extending instructions.
+
+However there is no offical rule on how to use those encoding space and the rule
+of assembly mnemonic, so this section is defining those rule to make it easier
+to management on upstream open source toolchain.
+
+### Vendor extension naming scheme
+
+According RISC-V ISA spec, non-standard extensions are named using a single `x`
+followed by an alphabetical name and an optional version number.
+
+For easier to identify and prevent naming conflict, vendor extension naming must
+start with vendor name, and that could be abbreviation.
+
+For example:
+- `XVentanaCondOps` from Ventana
+- `Xsfcflushdlone` from SiFive
+- `Xtheade` from T-head
+
+### Assembly mnemonic
+
+In order to avoid confusion between standard extension and other vendor
+extensions, vendor extension must prefixed with their own vendor prefix,
+e.g. `sf.` for SiFive, `vt.` for Ventana. and no registration required
```

Comment:
Let's require that vendors register their extensions by listing them in this document.
If something outrageous or unadvisable — e.g. because it has an offensive meaning in some geography (the fact that every class defined in NextStep starts with 'NS' still makes me cringe) or language — comes along, this will give the community a chance to raise their concerns).

Let's add a table for this in the document:

### List of vendor prefixes

Prefix  | Vendor | URL
:------ | :------ | :------
vt | Ventana Micro Systems | https://www.ventanamicro.com/
sf | SiFive | https://www.sifive.com/

Reply:
I suggest a primary prefix so that there is no concern with vendor prefixes polluting or reducing the standard mnemonic name space.  Such could be `x` to match the extension prefix, or something like `vnd` for "vendor".

- `x.vt.hcf`
- `xvt.hcf`
- `vnd.vt.hcf`

Reply:
> I suggest a primary prefix so that there is no concern with vendor prefixes polluting or reducing the standard mnemonic name space. Such could be `x` to match the extension prefix, or something like `vnd` for "vendor".
> 
> * `x.vt.hcf`
> * `xvt.hcf`
> * `vnd.vt.hcf`

@kito-cheng +1 for x.<vendor insn>, I think some user's interface more import like -march=*_xtheadc than internal name, we should make simple rule with assemble's name with existing rule "X"

Reply:
I am OK with `x.<vendor>.` prefix, @cmuellner @ptomsich @asb, any comment?

Reply:
I find this a bit unwieldy and would prefer to use the vendor-prefix only (i.e. vt/th/…) instead of further prefixing (one of the reasons being that I like my disassembly to be nicely aligned, which gets harder if some mnemonics become very long).  This may also get quite uncomfortable, if vendors then use the appropriate '.uw' postfix on some instruction (we'll end up with 3 '.' delimiters in the mnemonic...).  All of this amounts to "personal preference", though.

I can certainly see the other side of this as well, as there is precedent with the 'c.' prefix used for compressed.

So count me as weakly opposed (i.e., based on a gut-feeling only).

Reply:
Agreed w/Philipp.

Reply:
Maybe we could have a NOTE to mention the vendor prefix should be at least 2 char, that could prevent collision for most situation.

We (Toolchain maintainers, ISA folks, and Toolchain & runtime TG folks) are here as a gatekeeper to make sure the vendor prefix won't be too ambiguous with possible opcode.

And I guess I should express clearly: I don't have strong preference on `x.` or any other prefix, so basically I am OK if all other are OK with that, but yeah, but I feel vendor specific like `sf.`, `th.` or `vt.` prefix only is enough  and clearly, so I am also kind of weakly opposed too.

Reply:
I support a two-character minimum, so that the single-character prefix space is reserved for standard purposes.

Reply:
> I support a two-character minimum, so that the single-character prefix space is reserved for standard purposes.

That sounds sensible to me.